cpp/nacjd.cc

```diff
@@ -0,0 +1,371 @@
+        if (c == 'D')
+            state = 3;
+        else
+            state = 0;
+    } else if (state == 3) {
+        if (c == '"')
+            state = 4;
+        else
+            state = 0;
+    } else if (state == 4) {
+        if (c == ':')
+            state = 5;
+        else
+            state = 0;
+    } else if (state == 5) {
+        if (c == ',') {
+            ids_website.push_back(currentId_website);
+            currentId_website = "";
+            state = 0;
+        }
+        else
+            currentId_website += c;
+    }
+}
+
+
+bool downloadID(std::ofstream &json_new_titles, std::string &id, bool use_separator) {
+    const std::string DOWNLOAD_URL = "https://pcms.icpsr.umich.edu/pcms/api/1.0/studies/" + id +"/versions/V1/dats?page=https://www.icpsr.umich.edu/web/NACJD/studies/37644/export&user=";
+
+    Downloader downloader(DOWNLOAD_URL, Downloader::Params(), timeout * 1000);
+    if (downloader.anErrorOccurred()) {
+        LOG_WARNING("Error while downloading data for id " + id + ": " + downloader.getLastErrorMessage());
+        return false;
+    }
+     // Check for rate limiting and error status codes:
+    const HttpHeader http_header(downloader.getMessageHeader());
+    if (http_header.getStatusCode() == 429)
+        LOG_ERROR("we got rate limited!");
+    else if (http_header.getStatusCode() != 200) {
+        LOG_WARNING("NACJD returned HTTP status code " + std::to_string(http_header.getStatusCode()) + "! for NACJD id: " + id);
+        return false;
+    }
+    const std::string &json_document(downloader.getMessageBody());
+    std::shared_ptr<JSON::JSONNode> full_tree;
+    JSON::Parser parser(json_document);
+    if (not parser.parse(&full_tree)) {
+        LOG_ERROR("failed to parse JSON (" + parser.getErrorMessage() + "), download URL was: " + DOWNLOAD_URL);
+        return false;
+    }
+    const std::shared_ptr<const JSON::ObjectNode> top_node(JSON::JSONNode::CastToObjectNodeOrDie("full_tree", full_tree));
+    if (use_separator)
+        json_new_titles << "," << std::endl;
+    json_new_titles << top_node->toString() << std::endl;
+    return true;
+}
+
+
+void extractExistingIDsFromMarc(MARC::Reader *const marc_reader, std::set<std::string> &parsed_marc_ids) {
+    while (MARC::Record record = marc_reader->read()) {
+        std::string ppn = record.getControlNumber();
+        std::string id_035 = record.getFirstSubfieldValue("035", 'a');
+        if (StringUtil::Contains(ppn, "[ICPSR]")) {
+            StringUtil::ReplaceString("[ICPSR]", "", &ppn);
+            StringUtil::TrimWhite(&ppn);
+            parsed_marc_ids.emplace(ppn);
+        }
+        if (StringUtil::Contains(id_035, "[ICPSR]")) {
+            StringUtil::ReplaceString("[ICPSR]", "", &id_035);
+            StringUtil::TrimWhite(&id_035);
+            parsed_marc_ids.emplace(id_035);
+        }
+    }
+}
+
+
+void extractIDsFromWebsite(std::set<std::string> &parsed_marc_ids, unsigned * const number_of_new_ids) {
```

Comment:
Either make parsed_marc_ids const, or if that is not possible, make it a pointer.